text/0000-erfc-post-build-contexts.md

```diff
@@ -0,0 +1,396 @@
+[unresolved]: #unresolved-questions
+
+These are mostly intended to be resolved during the experimental
+feature. Many of these have strawman proposals -- unlike the rest of this RFC,
+these proposals have not been discussed as thoroughly. If folks feel like
+there's consensus on some of these we can move them into the main RFC.
+
+## Integration with doctests
+
+Documentation tests are somewhat special, in that they cannot easily be
+expressed as `TokenStream` manipulations. In the first instance, the
+right thing to do is probably to have an implicitly defined post-build
+context called `doctest` which is included in the post-build context set
+`test` by default (as proposed above).
+
+Another argument for punting on doctests is that they are intended to
+demonstrate code that the user of a library would write. They're there
+to document *how* something should be used, and it then makes somewhat
+less sense to have different "ways" of running them.
+
+## Translating existing cargo test flags
+
+Today, `cargo test` takes a number of flags such as `--lib`, `--test
+foo`, and `--doc`. As it would be a breaking change to change these,
+cargo should recognize them and map to the appropriate post-build
+contexts.
+
+Furthermore, `cargo test` lets you pick a single testing target via `--test`,
+and `cargo bench` via `--bench`. We'll need to create an agnostic flag
+for `cargo context` (we cannot use `--target` because it is already used for
+the target architecture, and `--test` is too specific for tests). `--post-build-target`
+is one rather verbose suggestion.
+
+We also need to settle on a command name, `cargo context` and `cargo post-build`
+don't quite capture what's going on.
+
+## Standardizing the output
+
+We should probably provide a crate with useful output formatters and
+stuff so that if test harnesses desire, they can use the same output
+formatting as a regular test. This also provides a centralized location
+to standardize things like json output and whatnot.
+
+## Namespacing
+
+Currently, two post-build contexts can both declare interest in the same
+attributes. How do we deal with collisions (e.g., most test crates will
+want the attribute `#[test]`). Do we namespace the attributes by the
+context name (e.g., `#[mytest::test]`)? Do we require them to be behind
+`#[cfg(mytest)]`?
+
+## Runtime dependencies and flags
+
+The code generated by the post-build context may itself have dependencies.
+Currently there's no way for the post-build context to specify this. One
+proposal is for the crate to specify  _runtime_ dependencies of the
+post-build context via:
+
+```toml
+[context-dependencies]
+libfuzzer-sys = ...
+```
+
+If a crate is currently running this post-build context, its
+dev-dependencies will be semver-merged with the post-build context's
+`context-dependencies`. However, this may not be strictly necessary.
+Custom derives have a similar problem and they solve it by just asking
+users to import the correct crate.
```

Comment:
Yeah. The specific motivation here is that libFuzzer is a dependency you do not want to get linked into other binaries by accident (it will break everything). But I think we can still include it in dev-dependencies as long as nobody tries to `extern crate` it.